### PR TITLE
[UNR-994] Clean up player actors after client disconnects

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
In native Unreal networking, when a client is closed, it closes the channel, which notifies the server that the client has disconnected, so the server can clean up that client's actors immediately.
To achieve that behavior in Spatial, there is now a field on the Heartbeat component that the client will set as it's shutting down. The server will clean up that client's connection upon receiving an update with that field.

#### Documentation
Release note

#### Primary reviewers
@m-samiec @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 1]
```diff
@@ -176,13 +174,12 @@ class USpatialReceiver : public UObject
 
 	AActor* FindSingletonActor(UClass* SingletonClass);
 
+	void OnHeartbeatComponentUpdate(Worker_ComponentUpdateOp& Op);
+
 public:
 	TMap<FUnrealObjectRef, TSet<FChannelObjectPair>> IncomingRefsMap;
 
 private:
-	template <typename T>
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]: Drive-by fix, this function has been removed a long time ago but declaration was still there.

File: Setup.bat [hunk 2]
```diff
@@ -76,21 +76,21 @@ call :MarkStartOfBlock "Create folders"
...
-    spatial package retrieve schema          standard_library                       %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\schema\standard_library.zip"
-    spatial package retrieve worker_sdk      c-dynamic-x86-msvc_md-win32            %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86-msvc_md-win32.zip"
-    spatial package retrieve worker_sdk      c-dynamic-x86_64-msvc_md-win32         %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86_64-msvc_md-win32.zip"
-    spatial package retrieve worker_sdk      c-dynamic-x86_64-gcc_libstdcpp-linux   %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86_64-gcc_libstdcpp-linux.zip"
-	spatial package retrieve worker_sdk      c-static-fullylinked-arm-clang_libcpp-ios  %PINNED_CORE_SDK_VERSION%   "%CORE_SDK_DIR%\worker_sdk\c-static-fullylinked-arm-clang_libcpp-ios.zip"
+    spatial package retrieve tools           schema_compiler-x86_64-win32               %PINNED_CORE_SDK_VERSION%   "%CORE_SDK_DIR%\tools\schema_compiler-x86_64-win32.zip"
```
Comment (improbable-valentyn) [inline Setup.bat]: No actual changes here, just adjusted the whitespace.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp [hunk 3]
```diff
@@ -115,6 +115,31 @@ void USpatialNetConnection::UpdateActorInterest(AActor* Actor)
...
+		Update.schema_type = Schema_CreateComponentUpdate(SpatialConstants::HEARTBEAT_COMPONENT_ID);
+		Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(Update.schema_type);
+
+		Schema_AddBool(ComponentObject, SpatialConstants::HEARTBEAT_CLIENT_HAS_QUIT_ID, true);
+
+		Cast<USpatialNetDriver>(Driver)->Connection->SendComponentUpdate(PlayerControllerEntity, &Update);
+	}
+	else
+	{
+		UE_LOG(LogSpatialNetConnection, Warning, TEXT("Quitting before Heartbeat component has been initialized: NetConnection %s"), *GetName());
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp]: Is this bad?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 4]
```diff
@@ -1618,3 +1611,50 @@ void USpatialReceiver::ResolveObjectReferences(FRepLayout& RepLayout, UObject* R
 		}
 	}
 }
+
+void USpatialReceiver::OnHeartbeatComponentUpdate(Worker_ComponentUpdateOp& Op)
+{
+	if (!NetDriver->IsServer())
+	{
+		// Clients can ignore Heartbeat component updates.
+		return;
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Not even sure this would ever get hit because clients can't see other PlayerControllers but it's probably fine to keep this.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp (hunk 3 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp]: This will likely mean that if there were actors spawned on the server for this client, they will stay there until they time out.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -195,14 +195,15 @@ void USpatialReceiver::HandlePlayerLifecycleAuthority(Worker_AuthorityChangeOp&
 		{
 			if (USpatialNetConnection* Connection = Cast<USpatialNetConnection>(PlayerController->GetNetConnection()))
 			{
+				EntityConnectionMap.Add(Op.entity_id, Connection);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I feel like the name of this map is a bit misleading. Since the PlayerController is checked out, we have created a connection for it, so if I naively saw this map, I would expect the player controller and connection to be in here even if I don't have authority. Potentially rename or maybe change logic to check for authority later?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 6]
```diff
@@ -855,10 +856,7 @@ void USpatialReceiver::OnComponentUpdate(Worker_ComponentUpdateOp& Op)
 #endif // WITH_EDITOR
 		return;
 	case SpatialConstants::HEARTBEAT_COMPONENT_ID:
-		if (HeartbeatDelegate* UpdateDelegate = HeartbeatDelegates.Find(Op.entity_id))
-		{
-			UpdateDelegate->ExecuteIfBound(Op);
-		}
+		OnHeartbeatComponentUpdate(Op);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Nice

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp [hunk 7]
```diff
@@ -115,6 +115,31 @@ void USpatialNetConnection::UpdateActorInterest(AActor* Actor)
 	}
 }
 
+void USpatialNetConnection::NotifyClientHasQuit()
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp]: As this is a client callable function, I'd rename to `ClientNotifyHasQuit`

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp]: But the name of the field is `client_has_quit`.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 8]
```diff
@@ -1618,3 +1614,50 @@ void USpatialReceiver::ResolveObjectReferences(FRepLayout& RepLayout, UObject* R
...
+
+		NetConnection->OnHeartbeat();
+	}
+
+	Schema_Object* FieldsObject = Schema_GetComponentUpdateFields(Op.update.schema_type);
+	if (Schema_GetBoolCount(FieldsObject, SpatialConstants::HEARTBEAT_CLIENT_HAS_QUIT_ID) > 0 &&
+		GetBoolFromSchema(FieldsObject, SpatialConstants::HEARTBEAT_CLIENT_HAS_QUIT_ID))
+	{
+		// Client has disconnected, let's clean up their connection.
+		NetConnection->CleanUp();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Should we delete the NetConnection here instead?

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: As far as I understand, `CleanUp` will remove the connection from the `NetDriver`, and it'll then be GC'd, so no additional actions are needed, unless you were referring to something else?

File: CHANGELOG.md [hunk 9]
```diff
@@ -4,6 +4,17 @@ All notable changes to the SpatialOS Game Development Kit for Unreal will be doc
 The format of this Changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased-`x.y.z`] - 2019-xx-xx
```
Comment (m-samiec) [inline CHANGELOG.md]: ❤️

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp (hunk 7 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp]: ClientNotifyClientHasQuit ? We should be following the function naming standard of prefixing function names if they're only intended to be called on client or server.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 8 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: No, just that CleanUp is called within the destructor.
